MovingBlocks/Terasology
# Display held items of remote players in multiplayer sessions.

## **Clients' local system for displaying held items of all remote players in multiplayer sessions.**

### Tackled issues: #3280

_Based on @Josharias 's firstPersonClientSystem introduced mid 2016, practically clones his changes._

[image summary](https://pasteboard.co/Hi09xqN.jpg)

### **Added java classes:**
**ThirdPersonRemoteClientSystem.java** - Essential class which glues all together. Handles displaying held items for all remote players in multiplayer session. Locally, registered for clients.
**RemotePersonHeldItemTransformComponent.java** - Component represents transformations of various held item types (used in hand,itemBase and blockItemBase prefabs)
**RemotePersonHeldItemMountPointComponent.java** - Component represents mounting point of characters for held items to attach to.
**ItemIsRemotelyHeldComponent.java** - This is used on the client side to track and clean up items that are no longer held by remote players, but are still location linked

### **Added prefabs:**
**remotePersonHeldItemMountPoint.prefab** - represents RemotePersonHeldItemMountPointComponent.java

### **Modified prefabs:**
**blockItemBase.prefab,itemBase.prefab,hand.prefab** - added remotePersonHeldItemTransform components to handle transformations of these items in game
**player.prefab** - added remotePersonHeldItemMountPoint component to attach all the held items to. Corresponds to same named .java

### **How to test and try out**
Host a server, run another instance, join the server, run around, change held items, pickup blocks, observe one another holding different types of held items (empty hand, blocks, weapons, torches)...

Run multiple instances of the game, host a multiplayer session, connect and disconnect players and watch their items. - ❗️  More than two players tends to crash often.


### **ToDo:**

### Outstanding before merging ❗️ 
 
- [ ] **Client sometimes logs in with a detached camera spawned at 0 0 0.**

- [ ] **Player's who happen to log in always see hands of other players or nothing until these other players perform change of the held item while the logged in player is in the game.**

- [ ] **Lingering previous held items**

- [ ] Occassional nullpointer crashes while: - simultaneously joining server, - on disconnecting of crashed player. Needs more testing and more eyes :) I've changed these mechanics so many times I can't see clearly.
